src/Nethermind/Nethermind.Optimism/CL/Decoding/FrameDecoder.cs

```diff
@@ -0,0 +1,121 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Optimism.CL.Decoding;
+
+public static class FrameDecoder
+{
+    // TODO: Evaluate if a custom `Enumerator` would be worth it
+    public static IEnumerable<Frame> DecodeFrames(Memory<byte> source)
+    {
+        while (source.Length != 0)
+        {
+            byte version = source.TakeAndMove(1).Span[0];
+            switch (version)
+            {
+                case 0:
+                    int bytesRead = Frame.FromBytes(source.Span, out Frame frame);
+                    yield return frame;
+                    source = source[bytesRead..];
+                    break;
+                default:
+                    throw new Exception($"Frame Decoder version {version} is not supported.");
+            }
+        }
+    }
+}
+
+/// <remarks>
+/// https://specs.optimism.io/protocol/derivation.html#frame-format
+/// </remarks>
+public readonly struct Frame : IEquatable<Frame>
```

Comment:
No need for `readonly` modifier on each property since the struct is `readonly` itself.